topic/machine-learning/llm-langchain/cratedb_rag_customer_support_langchain.ipynb

```diff
@@ -0,0 +1,424 @@
+    "\n",
+    "```shell\n",
+    "docker run --publish 4200:4200 --publish 5432:5432 --pull=always crate:latest -Cdiscovery.type=single-node\n",
+    "```\n",
+    "\n",
+    "## Setup\n",
+    "\n",
+    "Install required Python packages, and import Python modules."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "colab": {
+     "base_uri": "https://localhost:8080/"
+    },
+    "id": "RJyP1GEXNHUy",
+    "outputId": "9c62258f-f6a1-4578-ced4-40f15f586e9a",
+    "scrolled": false
+   },
+   "outputs": [],
+   "source": [
+    "!pip install -r requirements.txt"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "id": "VUNjBDrXNNoG"
+   },
+   "outputs": [],
+   "source": [
+    "import os\n",
+    "import re\n",
+    "\n",
+    "import openai\n",
+    "import pandas as pd\n",
+    "import sqlalchemy as sa\n",
+    "import warnings\n",
+    "from dotenv import load_dotenv\n",
+    "\n",
+    "from langchain.document_loaders.csv_loader import CSVLoader\n",
+    "from langchain_openai import OpenAIEmbeddings\n",
+    "from langchain.docstore.document import Document\n",
+    "from langchain.vectorstores import CrateDBVectorSearch\n",
+    "\n",
+    "warnings.filterwarnings('ignore')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Configure database settings"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This notebook will connect to a CrateDB server instance running on localhost. You can start a sandbox instance on your workstation by running [CrateDB using Docker]. Alternatively, you can also connect to a cluster running on [CrateDB Cloud].\n",
+    "\n",
+    "[CrateDB Cloud]: https://console.cratedb.cloud/\n",
+    "[CrateDB using Docker]: https://crate.io/docs/crate/tutorials/en/latest/basic/index.html#docker."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Define the connection string to running CrateDB instance.\n",
+    "CONNECTION_STRING = os.environ.get(\n",
+    "    \"CRATEDB_CONNECTION_STRING\",\n",
+    "    \"crate://crate@localhost/\",\n",
+    ")\n",
+    "\n",
+    "# Define the store collection to use for this notebook session.\n",
+    "COLLECTION_NAME = \"customer_data\""
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Configure OpenAI"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In this example you need to have an API key from OpenAI. This is typically done by creating an account on OpenAI's website and accessing the API section, where you can generate a new key."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "load_dotenv()\n",
+    "\n",
+    "openai_api_key = os.getenv('OPENAI_API_KEY')"
```

Comment:
Following @ckurze's suggestion, please also use the same code as in the other notebook here. Thanks.